pkg/client/cache/expiration_cache.go

```diff
@@ -0,0 +1,178 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"fmt"
+	"time"
+)
+
+// ExpirationCache implements the store interface
+//	1. All entries are automatically time stamped on insert
+//		a. Thekey is computed based off the original item/keyFunc
+//		b. The value inserted under that key is the timestamped item
+//	2. TTL expiration happens lazily on read, via List/Get
+//	3. Time-stamps are stripped off unexpired entries before return
+type ExpirationCache struct {
+	cacheStorage     ThreadSafeStore
+	keyFunc          KeyFunc
+	expirationPolicy ExpirationPolicy
+}
+
+// ExpirationPolicy dictates when an object expires. Currently only abstracted out
+// so unittests don't rely on the system clock.
+type ExpirationPolicy interface {
+	IsExpired(obj interface{}) bool
+}
+
+// TTLPolicy implements a ttl based ExpirationPolicy.
+type TTLPolicy struct {
+	//	 >0: Expire entries with an age > ttl
+	//	<=0: Don't expire any entry
+	ttl time.Duration
+}
+
+// IsExpired returns true if the given object is older than the ttl, or it can't
+// determine its age.
+func (p *TTLPolicy) IsExpired(obj interface{}) bool {
+	if timeStampedEntry, ok := obj.(*TimeStampedEntry); !ok {
+		return true
+	} else {
+		return p.ttl <= 0 || time.Since(timeStampedEntry.timeStamp) > p.ttl
+	}
+}
+
+// TimeStampedEntry is the only type allowed in a ExpirationCache.
+type TimeStampedEntry struct {
+	obj       interface{}
+	timeStamp time.Time
+}
+
+// getTimeStampedEntry returnes the TimeStampedEntry stored under the given key.
+func (c *ExpirationCache) getTimeStampedEntry(key string) (*TimeStampedEntry, bool) {
+	item, exists := c.cacheStorage.Get(key)
+	// TODO: Check the cast instead
+	if exists {
+		return item.(*TimeStampedEntry), exists
+	}
+	return nil, false
+}
+
+// getOrExpire retrieves the object from the TimeStampedEntry iff it hasn't
+// already expired. It kicks-off a go routine to delete expired objects from
+// the store and sets exists=false.
+func (c *ExpirationCache) getOrExpire(key string) (interface{}, bool) {
+	timeStampedItem, exists := c.getTimeStampedEntry(key)
+	if !exists {
+		return nil, false
+	}
+	if c.expirationPolicy.IsExpired(timeStampedItem) {
+		go func() { c.cacheStorage.Delete(key) }()
+		return nil, false
+	}
+	return timeStampedItem.obj, true
+}
+
+// GetByKey returns the item stored under the key, or sets exists=false.
+func (c *ExpirationCache) GetByKey(key string) (interface{}, bool, error) {
+	obj, exists := c.getOrExpire(key)
+	return obj, exists, nil
+}
+
+// Get returns unexpired items. It purges the cache of expired items in the
+// process.
+func (c *ExpirationCache) Get(obj interface{}) (interface{}, bool, error) {
+	key, err := c.keyFunc(obj)
+	if err != nil {
+		return nil, false, err
```

Comment:
If you could grab the commits from #5473 (except for the 'Add a "delta fifo", which queues up object changes' commit, which has a small chance of still changing), then there's a useful error type you could return here instead of just err.